lewis/adapters/modbus.py

```diff
@@ -0,0 +1,569 @@
+        subject to change.
+
+        .. sourcecode:: Python
+
+            di = ModbusDataBank.create_basic(False, 0x1000, 0x1FFF)
+
+        :param config: dict containing configuration
+        """
+        self._data = config['data']
+        self._start_addr = config['start_addr']
+
+    @classmethod
+    def create_basic(cls, default_value=0, start_addr=0x0000, last_addr=0xFFFF):
+        """
+        Create a basic ModbusDataBank instance.
+
+        This type of DataBank simply serves as a memory space for Modbus requests to read from and
+        write to. It does not support binding addresses to attributes or functions of the device
+        or interface.
+
+        :param default_value: Value to initialize memory with
+        :param start_addr: First valid address
+        :param last_addr: Last valid address
+        :return: ModbusDataBank instance with specified configuration
+        """
+        return cls({'start_addr': start_addr,
+                    'data': [default_value] * (last_addr - start_addr + 1)})
+
+    def get(self, addr, count):
+        """
+        Read list of ``count`` values at ``addr`` memory location in DataBank.
+
+        :param addr: Address to read from
+        :param count: Number of entries to retrieve
+        :return: list of entry values
+        :except IndexError: Raised if address range falls outside valid range
+        """
+        addr -= self._start_addr
+        data = self._data[addr:addr+count]
+        if len(data) != count:
+            addr += self._start_addr
+            raise IndexError("Invalid address range [{:#06x} - {:#06x}]"
+                             .format(addr, addr+count))
+        return data
+
+    def set(self, addr, values):
+        """
+        Write list ``values`` to ``addr`` memory location in DataBank.
+
+        :param addr: Address to write to
+        :param values: list of values to write
+        :except IndexError: Raised if address range falls outside valid range
+        """
+        addr -= self._start_addr
+        end = addr + len(values)
+        if not 0 <= addr <= end <= len(self._data):
+            addr += self._start_addr
+            raise IndexError("Invalid address range [{:#06x} - {:#06x}]"
+                             .format(addr, addr+len(values)))
+        self._data[addr:end] = values
+
+
+class ModbusDataStore(object):
+    """Convenience struct to hold the four types of DataBanks in Modbus"""
+    def __init__(self, di=None, co=None, ir=None, hr=None):
+        self.di = di
+        self.co = co
+        self.ir = ir
+        self.hr = hr
+
+
+class MBEX(object):
+    """Modbus standard exception codes"""
+    ILLEGAL_FUNCTION = 0x01
+    DATA_ADDRESS = 0x02
+    DATA_VALUE = 0x03
+    SLAVE_DEVICE_FAILURE = 0x04
+    ACKNOWLEDGE = 0x05
+    SLAVE_DEVICE_BUSY = 0x06
+    MEMORY_PARITY_ERROR = 0x08
+    GATEWAY_PATH_UNAVAILABLE = 0x0A
+    GATEWAY_TARGET_DEVICE_FAILED_TO_RESPOND = 0x0B
+
+
+class ModbusTCPFrame(object):
```

Comment:
You already mentioned that we need to refactor the Modbus and Stream infrastructure in order to minimize code duplication. Maybe TCP Stream could profit from something similar to this, although it might be harder to do this for "free format protocols". Something to think about for the next release.